subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/TaskContainerEvalTest.kt

```diff
@@ -0,0 +1,423 @@
+            val task: TaskNameAndType,
+            val descriptionMatcher: Matcher<in String?>
+        )
+
+        fun taskAssertion(
+            name: String,
+            type: KClass<out Task> = Task::class,
+            descriptionMatcher: Matcher<in String?> = nullValue()
+        ) =
+            TaskAssertion(TaskNameAndType(name, type), descriptionMatcher)
+
+        val preRegisteredTasks = listOf(
+            TaskNameAndType("foo"),
+            TaskNameAndType("bar", Copy::class),
+            TaskNameAndType("bat", Copy::class),
+            TaskNameAndType("pipistrelle", Copy::class)
+        )
+
+        val tasksConfigurationAssertions = listOf(
+            taskAssertion("foo"),
+            taskAssertion("bar", Copy::class, equalTo("null!!!")),
+            taskAssertion("bazar"),
+            taskAssertion("cathedral", Copy::class),
+            taskAssertion("cabin", Copy::class),
+            taskAssertion("castle", Task::class, equalTo("null!")),
+            taskAssertion("valley", Copy::class, equalTo("null!")),
+            taskAssertion("hill", Copy::class, equalTo("null!")),
+            taskAssertion("bat"),
+            taskAssertion("pipistrelle", Copy::class, equalTo("null!!!")),
+            taskAssertion("yate"),
+            taskAssertion("quartern", Copy::class),
+            taskAssertion("veduta", Copy::class),
+            taskAssertion("vansire", Task::class, equalTo("null!")),
+            taskAssertion("koto", Copy::class, equalTo("null!")),
+            taskAssertion("diptote", Copy::class, equalTo("null!"))
+        )
+    }
+
+    private
+    fun assertTasksConfiguration(
+        name: String,
+        before: Project.() -> Unit = {},
+        script: String,
+        tasksAssertions: List<TaskAssertion> = tasksConfigurationAssertions
+    ) {
+        newFolder(name).newProjectBuilderProject().run {
+
+            preRegisteredTasks.forEach {
+                tasks.register(it.name, it.type.java)
+            }
+
+            before()
+
+            eval(
+                script,
+                scriptCompilationClassPath = testCompilationClassPath + customInstallationGradleApiExtensionsClasspath
+            )
+
+            tasksAssertions.forEach { taskAssertion ->
+                taskAssertion.run {
+                    assertThat(
+                        "${task.name}: ${task.type.simpleName}",
+                        tasks.getByName(task.name, task.type).description,
+                        descriptionMatcher
+                    )
+                }
+            }
+        }
+    }
+
+    @Test
+    fun `polymorphic named domain object container api`() {
+
+        assertTasksConfiguration("api", script = """
+
+            val t1: Task = tasks.getByName("foo")
+            val t2: Task = tasks.getByName("foo", Task::class)
+            val t3: Task = tasks.getByName<Task>("foo")
+
+            val t4: Task = tasks.getByName("bar") {
+                description += "!"
```

Comment:
:x: it would be better to differentiate the strings added to the description so the proper ordering of execution is also tested.